File: src/main/java/me/helloc/iot/tunnel/MqttBufferedSubscriber.java
package me.helloc.iot.tunnel;

import org.eclipse.paho.client.mqttv3.IMqttActionListener;
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.IMqttToken;
import org.eclipse.paho.client.mqttv3.MqttAsyncClient;
import org.eclipse.paho.client.mqttv3.MqttCallbackExtended;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.function.Supplier;

/**
 * Manages MQTT connection with automatic reconnection and subscription restoration.
 */
public class MqttBufferedSubscriber {

    public interface ConnectionListener {
        void onConnected();
        void onConnectionLost(Throwable cause);
        void onDisconnected();
    }

    public interface MessageListener {
        void onMessageReceived(String topic, String message);
        void onBufferedBefore(String topic, String message);
        void onBufferedAfter(String topic, String message);
    }

    private final String brokerUrl;
    private final String clientId;
    private final String[] topics;
    private final MqttConnectOptions options;
    private final ScheduledExecutorService scheduler;
    public final MessageBuffer messageBuffer;
    private final Supplier<MqttAsyncClient> clientSupplier;
    private final MqttAsyncClient client;
    private final CopyOnWriteArrayList<ConnectionListener> listeners = new CopyOnWriteArrayList<>();
    private final CopyOnWriteArrayList<MessageListener> messageListeners = new CopyOnWriteArrayList<>();
    private final long initialDelay;
    private final long maxDelay;
    private final int qos;
    private volatile long currentDelay;

    private MqttBufferedSubscriber(Builder builder) {
        this.brokerUrl = Objects.requireNonNull(builder.brokerUrl, "brokerUrl");
        this.clientId = builder.clientId != null ? builder.clientId : MqttAsyncClient.generateClientId();
        this.topics = builder.topics.toArray(new String[0]);
        this.options = builder.options;
        this.scheduler = builder.scheduler != null ? builder.scheduler : Executors.newSingleThreadScheduledExecutor();
        this.messageBuffer = builder.messageBuffer;
        this.clientSupplier = builder.clientSupplier != null ? builder.clientSupplier : () -> {
            try {
                return new MqttAsyncClient(brokerUrl, clientId);
            } catch (MqttException e) {
                throw new RuntimeException(e);
            }
        };
        this.client = clientSupplier.get();
        this.client.setCallback(new InternalCallback());
        this.initialDelay = builder.initialDelay;
        this.maxDelay = builder.maxDelay;
        this.qos = builder.qos;
        this.currentDelay = initialDelay;
    }

    public void addListener(ConnectionListener listener) {
        listeners.add(listener);
    }

    public void removeListener(ConnectionListener listener) {
        listeners.remove(listener);
    }

    public void addMessageListener(MessageListener listener) {
        messageListeners.add(listener);
    }

    public void removeMessageListener(MessageListener listener) {
        messageListeners.remove(listener);
    }

    public void connect() {
        try {
            client.connect(options, null, new IMqttActionListener() {
                @Override
                public void onSuccess(IMqttToken asyncActionToken) {
                    currentDelay = initialDelay;
                    subscribeAll();
                    for (ConnectionListener l : listeners) {
                        l.onConnected();
                    }
                }

                @Override
                public void onFailure(IMqttToken asyncActionToken, Throwable exception) {
                    scheduleReconnect();
                }
            });
        } catch (MqttException e) {
            scheduleReconnect();
        }
    }

    public void disconnect() {
        scheduler.shutdownNow();
        try {
            client.disconnect();
            for (ConnectionListener l : listeners) {
                l.onDisconnected();
            }
        } catch (MqttException ignore) {
        }
    }

    private void scheduleReconnect() {
        scheduler.schedule(this::connect, currentDelay, TimeUnit.MILLISECONDS);
        currentDelay = Math.min(currentDelay * 2, maxDelay);
    }

    private void subscribeAll() {
        for (String topic : topics) {
            try {
                client.subscribe(topic, qos);
            } catch (MqttException ignore) {
            }
        }
    }

    private class InternalCallback implements MqttCallbackExtended {
        @Override
        public void connectComplete(boolean reconnect, String serverURI) {
            if (reconnect) {
                subscribeAll();
                for (ConnectionListener l : listeners) {
                    l.onConnected();
                }
            }
        }

        @Override
        public void connectionLost(Throwable cause) {
            if (cause != null) {
                for (ConnectionListener l : listeners) {
                    l.onConnectionLost(cause);
                }
            }
            scheduleReconnect();
        }

        @Override
        public void messageArrived(String topic, MqttMessage message) {
            if (topic != null && message != null) {
                String payload = new String(message.getPayload());
                for (MessageListener l : messageListeners) {
                    l.onMessageReceived(topic, payload);
                }
                for (MessageListener l : messageListeners) {
                    l.onBufferedBefore(topic, payload);
                }
                messageBuffer.add(topic, payload);
                for (MessageListener l : messageListeners) {
                    l.onBufferedAfter(topic, payload);
                }
            }
        }

        @Override
        public void deliveryComplete(IMqttDeliveryToken token) {
        }
    }

    public static class Builder {
        private String brokerUrl;
        private String clientId;
        private final List<String> topics = new ArrayList<>();
        private MqttConnectOptions options = new MqttConnectOptions();
        private ScheduledExecutorService scheduler;
        private Supplier<MqttAsyncClient> clientSupplier;
        private long initialDelay = 1000;
        private long maxDelay = 60000;
        private MessageBuffer messageBuffer = new InMemoryMessageBuffer();
        private int qos = 1;

        public Builder brokerUrl(String brokerUrl) { this.brokerUrl = brokerUrl; return this; }
        public Builder clientId(String clientId) { this.clientId = clientId; return this; }
        public Builder addTopic(String topic) { this.topics.add(topic); return this; }
        public Builder options(MqttConnectOptions options) { this.options = options; return this; }
        public Builder scheduler(ScheduledExecutorService scheduler) { this.scheduler = scheduler; return this; }
        public Builder clientSupplier(Supplier<MqttAsyncClient> supplier) { this.clientSupplier = supplier; return this; }
        public Builder initialDelay(long delay) { this.initialDelay = delay; return this; }
        public Builder maxDelay(long delay) { this.maxDelay = delay; return this; }
        public Builder messageBuffer(MessageBuffer buffer) { this.messageBuffer = buffer; return this; }
        public Builder qos(int qos) {
            if (qos < 0 || qos > 2) throw new IllegalArgumentException("QoS must be between 0 and 2");
            this.qos = qos;
            return this;
        }

        public MqttBufferedSubscriber build() {
            if (brokerUrl == null || brokerUrl.isEmpty()) {
                throw new IllegalArgumentException("brokerUrl must not be null or empty");
            }
            return new MqttBufferedSubscriber(this);
        }
    }

    public static Builder builder() {
        return new Builder();
    }

    /**
     * Default adapter implementation of ConnectionListener.
     * Provides empty implementations of all methods.
     */
    public static abstract class DefaultConnectionListener implements ConnectionListener {
        @Override
        public void onConnected() {
        }

        @Override
        public void onConnectionLost(Throwable cause) {
        }

        @Override
        public void onDisconnected() {
        }
    }

    /**
     * Default adapter implementation of MessageListener.
     * Provides empty implementations of all methods.
     */
    public static abstract class DefaultMessageListener implements MessageListener {
        @Override
        public void onMessageReceived(String topic, String message) {
        }

        @Override
        public void onBufferedBefore(String topic, String message) {
        }

        @Override
        public void onBufferedAfter(String topic, String message) {
        }
    }

    /**
     * Default adapter implementation of both ConnectionListener and MessageListener.
     * Provides empty implementations of all methods from both interfaces.
     */
    public static abstract class DefaultMqttListener implements ConnectionListener, MessageListener {
        @Override
        public void onConnected() {
        }

        @Override
        public void onConnectionLost(Throwable cause) {
        }

        @Override
        public void onDisconnected() {
        }

        @Override
        public void onMessageReceived(String topic, String message) {
        }

        @Override
        public void onBufferedBefore(String topic, String message) {
        }

        @Override
        public void onBufferedAfter(String topic, String message) {
        }
    }
}

File: src/main/kotlin/me/helloc/iot/tunnel/MqttBufferedSubscriber.kt
package me.helloc.iot.tunnel

import org.eclipse.paho.client.mqttv3.IMqttActionListener
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken
import org.eclipse.paho.client.mqttv3.IMqttToken
import org.eclipse.paho.client.mqttv3.MqttAsyncClient
import org.eclipse.paho.client.mqttv3.MqttCallbackExtended
import org.eclipse.paho.client.mqttv3.MqttConnectOptions
import org.eclipse.paho.client.mqttv3.MqttException
import org.eclipse.paho.client.mqttv3.MqttMessage
import me.helloc.iot.tunnel.MessageBuffer
import me.helloc.iot.tunnel.MessageBufferFactory
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit

/**
 * Manages MQTT connection with automatic reconnection and subscription restoration.
 */
class MqttBufferedSubscriber private constructor(builder: Builder) {

    interface ConnectionListener {
        fun onConnected()
        fun onConnectionLost(cause: Throwable)
        fun onDisconnected()
    }

    interface MessageListener {
        fun onMessageReceived(topic: String, message: String)
        fun onBufferedBefore(topic: String, message: String)
        fun onBufferedAfter(topic: String, message: String)
    }

    private val brokerUrl: String = builder.brokerUrl!!
    private val clientId: String = builder.clientId ?: MqttAsyncClient.generateClientId()
    private val topics: Array<String> = builder.topics.toTypedArray()
    private val options: MqttConnectOptions = builder.options
    private val scheduler: ScheduledExecutorService = builder.scheduler ?: Executors.newSingleThreadScheduledExecutor()
    val messageBuffer: MessageBuffer = builder.messageBuffer
    private val clientSupplier: () -> MqttAsyncClient = builder.clientSupplier ?: {
        MqttAsyncClient(brokerUrl, clientId)
    }
    private val client: MqttAsyncClient = clientSupplier.invoke().apply {
        setCallback(InternalCallback())
    }
    private val listeners = CopyOnWriteArrayList<ConnectionListener>()
    private val messageListeners = CopyOnWriteArrayList<MessageListener>()
    private val initialDelay: Long = builder.initialDelay
    private val maxDelay: Long = builder.maxDelay
    private val qos: Int = builder.qos
    @Volatile private var currentDelay: Long = initialDelay

    fun addListener(listener: ConnectionListener) {
        listeners.add(listener)
    }

    fun removeListener(listener: ConnectionListener) {
        listeners.remove(listener)
    }

    fun addMessageListener(listener: MessageListener) {
        messageListeners.add(listener)
    }

    fun removeMessageListener(listener: MessageListener) {
        messageListeners.remove(listener)
    }

    fun connect() {
        try {
            client.connect(options, null, object : IMqttActionListener {
                override fun onSuccess(asyncActionToken: IMqttToken?) {
                    currentDelay = initialDelay
                    subscribeAll()
                    listeners.forEach { it.onConnected() }
                }

                override fun onFailure(asyncActionToken: IMqttToken?, exception: Throwable?) {
                    scheduleReconnect()
                }
            })
        } catch (e: MqttException) {
            scheduleReconnect()
        }
    }

    fun disconnect() {
        scheduler.shutdownNow()
        try {
            client.disconnect()
            listeners.forEach { it.onDisconnected() }
        } catch (_: MqttException) {
        }
    }

    private fun scheduleReconnect() {
        scheduler.schedule({ connect() }, currentDelay, TimeUnit.MILLISECONDS)
        currentDelay = (currentDelay * 2).coerceAtMost(maxDelay)
    }

    private fun subscribeAll() {
        for (topic in topics) {
            try {
                client.subscribe(topic, qos)
            } catch (_: MqttException) {
            }
        }
    }

    private inner class InternalCallback : MqttCallbackExtended {
        override fun connectComplete(reconnect: Boolean, serverURI: String?) {
            if (reconnect) {
                subscribeAll()
                listeners.forEach { it.onConnected() }
            }
        }

        override fun connectionLost(cause: Throwable?) {
            if (cause != null) listeners.forEach { it.onConnectionLost(cause) }
            scheduleReconnect()
        }

        override fun messageArrived(topic: String?, message: MqttMessage?) {
            if (topic != null && message != null) {
                val payload = String(message.payload)
                messageListeners.forEach { it.onMessageReceived(topic, payload) }
                messageListeners.forEach { it.onBufferedBefore(topic, payload) }
                messageBuffer.add(topic, payload)
                messageListeners.forEach { it.onBufferedAfter(topic, payload) }
            }
        }

        override fun deliveryComplete(token: IMqttDeliveryToken?) {
        }
    }

    class Builder {
        var brokerUrl: String? = null
            private set
        var clientId: String? = null
            private set
        internal val topics = mutableListOf<String>()
        var options: MqttConnectOptions = MqttConnectOptions()
            private set
        var scheduler: ScheduledExecutorService? = null
            private set
        var clientSupplier: (() -> MqttAsyncClient)? = null
            private set
        var initialDelay: Long = 1000
            private set
        var maxDelay: Long = 60000
            private set
        var messageBuffer: MessageBuffer = MessageBufferFactory.fromConfig()
            private set
        var qos: Int = 1
            private set

        fun brokerUrl(brokerUrl: String) = apply { this.brokerUrl = brokerUrl }
        fun clientId(clientId: String) = apply { this.clientId = clientId }
        fun addTopic(topic: String) = apply { this.topics.add(topic) }
        fun options(options: MqttConnectOptions) = apply { this.options = options }
        fun scheduler(scheduler: ScheduledExecutorService) = apply { this.scheduler = scheduler }
        fun clientSupplier(supplier: () -> MqttAsyncClient) = apply { this.clientSupplier = supplier }
        fun initialDelay(delay: Long) = apply { this.initialDelay = delay }
        fun maxDelay(delay: Long) = apply { this.maxDelay = delay }
        fun messageBuffer(buffer: MessageBuffer) = apply { this.messageBuffer = buffer }
        fun qos(qos: Int) = apply {
            require(qos in 0..2) { "QoS must be between 0 and 2" }
            this.qos = qos
        }

        fun build(): MqttBufferedSubscriber {
            require(!brokerUrl.isNullOrEmpty()) { "brokerUrl" }
            return MqttBufferedSubscriber(this)
        }
    }

    companion object {
        fun builder() = Builder()
    }
}
